### PR TITLE
Save `Processor`

### PR DESCRIPTION
**Pull request by ydshieh**
# What does this PR do?

Save `Processor`.

### CODE REVIEW

File: src/transformers/processing_utils.py
Comment (ydshieh) [inline src/transformers/processing_utils.py:475]: I would like to hear from the core maintainers to see if there is a better approach for this situation 🙏

File: src/transformers/processing_utils.py
Comment (ydshieh) [inline src/transformers/processing_utils.py:114]: Would like to hear from @amyeroberts  for this part as they have the experience for `ImageProcessingMixin`.

#### Question in short:

Is it OK we only save the attributes that are also presented in `__init__`'s argument?

#### Details 

Basically, for `ImageProcessingMixin`, it seems we have the correspondence between `the attributes of self` and the arguments presented in `self.__init__`.

However, for `Processor`, this is not the case.

Example 1:

```python
class FuyuProcessor(ProcessorMixin):

    def __init__(self, image_processor, tokenizer):
        ...
        ...
        self.max_tokens_to_generate = 10
        self.max_position_embeddings = 16384
```
For `FuyuProcessor`, we could add those attributes to `__init__` arguments.

Example 2:

```python
class Kosmos2Processor(ProcessorMixin):

    def __init__(self, image_processor, tokenizer, num_patch_index_tokens=1024):
        tokenizer.return_token_type_ids = False

        self.eod_token = "</doc>"
        self.boi_token = "<image>"
        self.tag_tokens = [....]

        self.num_patch_index_tokens = num_patch_index_tokens
```
Some of the (token) attributes have to be there (used in other methods) and others are kept as information. But **it doesn't make much sense to put them in the arguments as their values are fixed**.

**Review by ydshieh — COMMENTED** [review]
Almost done, but I have 2 design questions 🙏 .

File: src/transformers/processing_utils.py
Comment (amyeroberts) [reply src/transformers/processing_utils.py:114]: I think it's better to just save out what the user can modify themselves i.e. what's in the `__init__`. If I looked at a config file e.g.:

```json
{
    "a": 1,
    "b": 2,
    "c": 3,
}
```

I would assume that I could modify `a`, `b` and `c`. However if e.g. `c` is hard-coded in the config then doing:

```py
config = FooConfig.from_pretrained(checkpoint, c=6)
```

Would result in unexpected behaviour and the only way for users to know what they can modify is looking into the code which is something we want to avoid.

File: src/transformers/processing_utils.py
Comment (amyeroberts) [reply src/transformers/processing_utils.py:475]: Tbh, I like this handling. It's backwards compatible and only sets a fallback behaviour if a specific error message is raised.

**Comment by ydshieh** [comment]
Thanks for giving your opinion 👍  Ready for the review.

There is no test for this new feature so far, as we don't have a common processor test file yet.

There is #27720 however.

**Comment by amyeroberts** [comment]
@ydshieh Thanks for all the work in this PR! 

I'm wondering if we should hold off on saving processor configs out until we refactor the processors to standardise their inputs and outputs. Otherwise we're going to introduce config files and values which we'll have to maintain backward compatibility with cc @molbap 

Would not merging be blocking anything else?

**Comment by ydshieh** [comment]
Good point! It makes a lot of sense, thank you.

#27718 is waiting this PR I believe.

But let me double check:

> refactor the processors to standardise their inputs and outputs

this sounds to me that we are talking about the arguments in a process's `call` method, in particular `text`, `images` etc (the most important ones). However, saving a config file for a processor only saves `__init__` argument, which is not about input/outpu of processor.

Do I miss anything in your message?

**Comment by amyeroberts** [comment]
@ydshieh Nope - you're completely right - I forgot it's just the processing classes passed to the processor init. 

Part of the future plan is to potentially have modality specific processors e.g. TextImageProcessor to make it more explicit in e.g. pipelines how inputs and outputs should be handled. I'm guessing this won't be an issue here as the e.g. TextImageProcessor acts more like the Auto classes and can use model specific configs to construct themselves?

**Comment by ydshieh** [comment]
If we ever introduce `TextImageProcessor` and/or `TextAudioProcessor` etc, I think they are not exactly same as auto classes but instead an abstraction one level above the explicit classes.

- auto: allow to load any checkpoint: `AutoProcessor.from_pretrained(".....")` should work
- explicit class: only allow to load a checkpoint of a specific model type: `CLIPProcessor.from_pretrained(".....")` should work for checkpoint of a CLIP processor but not others (well, we allow it by give warnings)
- `TextImageProcessor`: they should not to be used as `TextImageProcessor.from_pretrained()`, just like we don't use `ProcessorMixin.from_pretrained()`

------------------

But back to the question

>  I'm guessing this won't be an issue here and can use model specific configs to construct themselves

If we ever allow to use `TextImageProcessor.from_pretrained()` (which I am against) and behave like `AutoProcessor`, then under the hood it still try to find the actual type (class) and use that type + the config file to construct  the object of that type.

**So I don't see any issue regarding such `TextImageProcessor` with the change in this PR 🤗**

**Comment by ydshieh** [comment]
Well, after a 2nd though, I think it's better to add the processor common testing for this PR to avoid the situation where we won't be able to fix in the future once the new processor config file being saved have some issues.

File: src/transformers/processing_utils.py
Comment (ydshieh) [inline src/transformers/processing_utils.py:116]: We have `image_processor_type` and `feature_extractor_type` being used, but `processor_class` is used in [src/transformers/models/auto/processing_auto.py](https://github.com/huggingface/transformers/blob/main/src/transformers/models/auto/processing_auto.py)

Note that `processor_class` is checked if being in the [image processor/feature extraoctor] file or if in the model config file, not the new [processor] file (yet).

@amyeroberts I think it's best to keep `processor_class` in the new [processor] file instead of `processor_type`. Do you have any opinion here?

File: src/transformers/processing_utils.py
Comment (amyeroberts) [inline src/transformers/processing_utils.py:456]: Could we have a logger.debug or logger.info for `e` to make sure it's logged in case of debugging something?

**Review by amyeroberts — APPROVED** [review]
Great work - thanks for adding!

File: src/transformers/processing_utils.py
Comment (ydshieh) [reply src/transformers/processing_utils.py:456]: Sure @amyeroberts . BTW, the PR is still in draft mode, as I haven't added some logic for `AutoProcessor`. (It should be a small change though)

Comment (amyeroberts) [reply src/transformers/processing_utils.py:456]: I'll give a final once-over review once moved out of drafts :)

File: tests/test_processing_common.py
Comment (ArthurZucker) [inline tests/test_processing_common.py:1]: We need to add tests to make sure we only call the hub once, make sure kwargs passed to the processor class are correctly forwarded to the __init__ of the tokenizer and the feature extractor (specifically trust_remote_code, and additional kwargs that do not exist are not duplicated)

**Review by ArthurZucker — COMMENTED** [review]
feel free to ping me as well if you need a full review when not in draft mode!

File: tests/test_processing_common.py
Comment (ydshieh) [reply tests/test_processing_common.py:1]: > We need to add tests to make sure we only call the hub once

well, I found we only have this tests for torch/tf model, tokenizer and pipeline (in auto). Not sure why we don't have this for image/feature extractor/processor.

**Comment by ydshieh** [comment]
The PR is ready for review. Regarding [the comment](https://github.com/huggingface/transformers/pull/27761#discussion_r1448533919), see [my response](https://github.com/huggingface/transformers/pull/27761#discussion_r1448552231).

For [we only call the hub once](https://github.com/huggingface/transformers/pull/27761#discussion_r1448620683), as we don't have such tests for image processor/feature extractor/processor currently, and it's not critical, I think it's better to deal with them in a follow up PR.

(In fact, I don't know if it is true that we can have only one call to the Hub for processor - considering processor have 2 components)

File: tests/models/auto/test_processor_auto.py
Comment (ydshieh) [inline tests/models/auto/test_processor_auto.py:114]: test by using the "processor_class" in the (new) `processor_config.json`

File: tests/models/auto/test_processor_auto.py
Comment (ydshieh) [inline tests/models/auto/test_processor_auto.py:348]: test remote code of processor that has extra arguments (and saved in a `processor_config.json`)

File: tests/test_processing_common.py
Comment (ydshieh) [inline tests/test_processing_common.py:127]: test local code of processor that has extra arguments (and saved in a processor_config.json)

**Review by ydshieh — COMMENTED** [review]
a few comments

**Review by amyeroberts — APPROVED** [review]
Thanks for all the work on this - looks great!

File: src/transformers/processing_utils.py
Comment (ArthurZucker) [inline src/transformers/processing_utils.py:389]:
```suggestion
        # Update processor with kwargs if needed
        to_remove = []
        for keyin inspect.signature(self.__init__) :
            if key in kwargs:
                setattr(processor, key, kwargs.pop("key))
                to_remove.append(key)
```
should be simpler

File: src/transformers/processing_utils.py
Comment (ArthurZucker) [inline src/transformers/processing_utils.py:391]:
```suggestion
```

File: src/transformers/processing_utils.py
Comment (ArthurZucker) [inline src/transformers/processing_utils.py:466]: this is a bit brittle but alright

File: tests/models/auto/test_processor_auto.py
Comment (ArthurZucker) [inline tests/models/auto/test_processor_auto.py:104]:
```suggestion
            # drop `processor_class` in tokenizer config
```

File: tests/models/auto/test_processor_auto.py
Comment (ArthurZucker) [inline tests/models/auto/test_processor_auto.py:329]: that is nice, documentation regarding custom code should be updated in a follow up PR

File: tests/test_processing_common.py
Comment (ArthurZucker) [inline tests/test_processing_common.py:2]:
```suggestion
# Copyright 2024 The HuggingFace Inc. team. All rights reserved.
```

File: tests/test_processing_common.py
Comment (ArthurZucker) [inline tests/test_processing_common.py:60]: naming conventions are a bit new? 
Would be nice to keep some equivalence between the equivalent config / tokenizer / feature extractors tests but anyway.

File: tests/test_processing_common.py
Comment (ArthurZucker) [inline tests/test_processing_common.py:39]: missing as setup function for tempdirname that is pretty usefull

File: tests/test_processing_common.py
Comment (ArthurZucker) [inline tests/test_processing_common.py:94]: if only used in the next test, define in the next test

**Review by ArthurZucker — COMMENTED** [review]
Left a few comments

File: src/transformers/processing_utils.py
Comment (ydshieh) [reply src/transformers/processing_utils.py:389]: I would try to avoid assuming that any key in `inspect.signature(self.__init__)` would be an attribute in `processor`.
For `processor`, yes, that's what we should maintain, but I don't want to the code block here to `assume` this.

Second, if a key in `kwargs` that is not in `inspect.signature(self.__init__)`, it woon't be added to `to_remove` with the suggest change, which we should add it to `to_remove`.

File: src/transformers/processing_utils.py
Comment (ydshieh) [reply src/transformers/processing_utils.py:391]: This is copied from `src/transformers/image_processing_utils.py`. Since this is just `logger.info`, it shouldn't cause problem

File: src/transformers/processing_utils.py
Comment (ydshieh) [reply src/transformers/processing_utils.py:466]: Yeah, this is not ideal but I don't see better way.

File: tests/models/auto/test_processor_auto.py
Comment (ydshieh) [reply tests/models/auto/test_processor_auto.py:329]: ok

File: tests/test_processing_common.py
Comment (ydshieh) [reply tests/test_processing_common.py:60]: the difference is that `config / tokenizer / feature extractors` don't have components, but processor has.

File: tests/test_processing_common.py
Comment (ydshieh) [reply tests/test_processing_common.py:39]: setup function so far are in the subclasses (explicitly) of `ProcessorTesterMixin`. They have their own logic and are not in common.

Comment (ydshieh) [reply tests/test_processing_common.py:39]: If you are talking just `self.tmpdirname = tempfile.mkdtemp()`, yes, I can move that to `ProcessorTesterMixin`

File: tests/test_processing_common.py
Comment (ydshieh) [reply tests/test_processing_common.py:99]: It's too small at this moment to be worth in a separate file IMO.

File: tests/test_processing_common.py
Comment (ydshieh) [reply tests/test_processing_common.py:94]: I did tried it, but it complains that the class `MyProcessor` is not defined if it is defined as inner class under `ProcessorTest`. I am not sure the reason behind ...


> tests/test_processing_common.py::ProcessorTest::test_processor_to_json_string - NameError: name 'MyProcessor' is not define

File: src/transformers/processing_utils.py
Comment (ArthurZucker) [reply src/transformers/processing_utils.py:391]: for sure it's just we try to have less warnings but alright

File: tests/test_processing_common.py
Comment (ArthurZucker) [reply tests/test_processing_common.py:39]: yes was the only relevant one IMO

File: tests/test_processing_common.py
Comment (ArthurZucker) [reply tests/test_processing_common.py:94]: No worries then

File: tests/test_processing_common.py
Comment (ydshieh) [reply tests/test_processing_common.py:39]: OK, I will do it in a follow up PR that will update all processor test file to inherit from `ProcessorTesterMixin`.

Note: the current existing processor tests won't benefit from this new `ProcessorTesterMixin` --> only the new processors in the future that have extra arguments (to be saved) will

Comment (ArthurZucker) [reply tests/test_processing_common.py:39]: Ok

**Comment by ydshieh** [comment]
I guess we are all good and I can merge?

File: src/transformers/processing_utils.py
Comment (ArthurZucker) [reply src/transformers/processing_utils.py:389]: The code is overly compilcated to do something quite simple

**Comment by ArthurZucker** [comment]
Still think we can simplify the way you handle kwargs, but alright otherwise

File: src/transformers/processing_utils.py
Comment (ydshieh) [reply src/transformers/processing_utils.py:389]: I disagree this is is overly compilcated  compared to what has been suggested, which also introduce unnecessary assumption of  `inspect.signature(self.__init__)`  and `processor`'s attributes.

And actually that code block is that small and direct, I don't understand why it is complicated - it's almost trivial to understand.

I can change it to manipulate the keys of `kwargs` and `processor.to_dict()` as two sets, but there is no real benefit IMO, as it again has to `processor.to_dict()` as well as `set()` construct as dict's `key()` are not set immediately.

Comment (ydshieh) [reply src/transformers/processing_utils.py:389]: OK, let me try one thing

Comment (ydshieh) [reply src/transformers/processing_utils.py]: Is the last commit ok

```python
        for key in set(kwargs.keys()):
            if hasattr(processor, key):
                setattr(processor, key, kwargs.pop(key))
```

**Pull request merged**
